File: 10/rb/cathode_ray.rb
# mostly by ChatGPT 🙃

def run(path="../input.txt")
    file_data = File.read(path).strip.split("\n")
  
    values = get_values(file_data)
    sum_signal_strength = [20, 60, 100, 140, 180, 220].sum { |cycle| cycle * values[cycle-1] }
    puts sum_signal_strength
  
    make_crt(file_data)
  end
  
  def get_values(data)
    values, x = [0], 1
    data.each do |line|
      if line.start_with?("noop")
        values << x
      else
        line = line.split(" ")
        values << x
        x += line[1].to_i
        values << x
      end
    end
    values
  end
  
  def make_crt(data, line=40)
    values = get_values(data)
    crt = values.map.with_index { |index, i| (index - (i % line)).abs <= 1 ? '#' : '.' }
    (0...values.length.div(line)).each do |i|
      puts crt[i*line, (i+1)*line].join('')
    end
  end

File: 10/ts/cathode-ray.ts
function getValues(data) {
    let values = [0]
    let x = 1

    for (let i = 0; i < data.length; i++) {
        if (data[i].startsWith('noop')) {
            values.push(x);
        } else {
            let line = data[i].split(' ');
            values.push(x);
            x += parseInt(line[1]);
            values.push(x);
        }
    }
    return values;
}

function sumValues(values) {
    let total = 0;
    for (let i = 20; i <= 220; i+=40) {
        total += values[i-1] * i
    }
    return total;
}

function makeCRT(values, line) {
    let crt = values.map((index, i) =>
        Math.abs(index - (i % line)) <= 1 ? '#' : ' '
    );

    for (let i = 0; i < values.length / line; i++) {
        console.log(crt
            .slice(i * line, (i + 1) * line)
            .join('')
        );
    }
}

function run() {
    const fs = require('fs');
    const data = fs
        .readFileSync('../input.txt', 'utf-8')
        .split('\n');

    let values = getValues(data);
    console.log(sumValues(values));
    makeCRT(values, 40);
}

run()